app/gui/view/graph-editor/src/lib.rs

```diff
@@ -1539,6 +1625,69 @@ impl GraphEditorModel {
 }
 
 
+// === Add node ===
+impl GraphEditorModel {
+    pub fn add_node(&self) -> NodeId {
+        self.frp.add_node.emit(());
+        let (node_id, _) = self.frp.node_added.value();
+        node_id
+    }
+
+    pub fn add_node_below(&self, above: NodeId) -> NodeId {
+        let pos = self.find_free_place_under(above);
+        self.add_node_at(pos)
+    }
+
+    pub fn add_node_at(&self, pos: Vector2) -> NodeId {
+        let node_id = self.add_node();
+        self.frp.set_node_position((node_id, pos));
+        node_id
+    }
+
+    pub fn find_free_place_under(&self, node_above: NodeId) -> Vector2 {
+        let above_pos = self.node_position(node_above);
+        let y_gap = self.frp.default_y_gap_between_nodes.value();
+        let y_offset = y_gap + node::HEIGHT;
+        let starting_point = above_pos - Vector2(0.0, y_offset);
+        let direction = Vector2(-1.0, 0.0);
+        self.find_free_place_for_node(starting_point, direction).unwrap()
+    }
+
+    pub fn find_free_place_for_node(
+        &self,
+        starting_from: Vector2,
+        direction: Vector2,
+    ) -> Option<Vector2> {
+        let x_gap = self.frp.default_x_gap_between_nodes.value();
+        let y_gap = self.frp.default_y_gap_between_nodes.value();
+        // This is how much horizontal space we are looking for.
+        let min_spacing = self.frp.min_x_spacing_for_new_nodes.value();
+        let nodes = self.nodes.all.raw.borrow();
+        // The "occupied area" for given node consists of:
+        // - area taken by node view (obviously);
+        // - the minimum gap between nodes in all directions, so the new node won't be "glued" to
+        //   another;
+        // - the new node size measured from origin point at each direction accordingly: because
+        //   `find_free_place` looks for free place for the origin point, and we want to fit not
+        //   only the point, but the whole node.
+        let node_areas = nodes.values().map(|node| {
+            let position = node.position();
+            let left = position.x - x_gap - min_spacing;
+            let right = position.x + node.view.model.width() + x_gap;
+            let top = position.y + node::HEIGHT + y_gap;
+            let bottom = position.y - node::HEIGHT - y_gap;
+            OccupiedArea { x1: left, x2: right, y1: top, y2: bottom }
+        });
+        find_free_place(starting_from, direction, node_areas)
+    }
+
+    pub fn start_editing_new_node(&self, node_id: NodeId) {
+        self.frp.set_node_expression.emit(&(node_id, node::Expression::default()));
+        self.frp.edit_node.emit(&node_id);
+    }
```

Comment:
This function seems to be unused. Is it deliberate?

Reply:
No, it is not. Thanks for catching!